Zend/zend_inheritance.c

```diff
@@ -3513,23 +3513,32 @@ ZEND_API zend_class_entry *zend_do_link_class(zend_class_entry *ce, zend_string
 		}
 	}
 
+	size_t num_implementable_interfaces = 0;
 	if (ce->num_interfaces) {
 		for (i = 0; i < ce->num_interfaces; i++) {
 			zend_class_entry *iface = zend_fetch_class_by_name(
 				ce->interface_names[i].name, ce->interface_names[i].lc_name,
 				ZEND_FETCH_CLASS_INTERFACE |
-				ZEND_FETCH_CLASS_ALLOW_NEARLY_LINKED | ZEND_FETCH_CLASS_EXCEPTION);
+				ZEND_FETCH_CLASS_ALLOW_NEARLY_LINKED | ZEND_FETCH_CLASS_EXCEPTION |
+				(ce->interface_names[i].is_optional ? ZEND_FETCH_CLASS_SILENT : 0));
+
+			// Optional interfaces are skipped if they don't exist.
+			if (!iface && ce->interface_names[i].is_optional) {
+				continue;
+			}
+
 			if (!iface) {
 				check_unrecoverable_load_failure(ce);
 				free_alloca(traits_and_interfaces, use_heap);
 				return NULL;
 			}
-			traits_and_interfaces[ce->num_traits + i] = iface;
+			traits_and_interfaces[ce->num_traits + num_implementable_interfaces++] = iface;
 			if (iface) {
 				UPDATE_IS_CACHEABLE(iface);
 			}
 		}
 	}
+	ce->num_interfaces = num_implementable_interfaces;
```

Comment:
This line is the reason your tests are segfaulting, this could be an immutable class entry. You may only modify immutable and non cached `ce`'s. Move this line to after the lazy class loading code in the zend_try down below.
e.g. right before `if (CG(unlinked_uses)) {` (probably).

Reply:
I also wonder what caching semantics will play in the following scenario:
1. You do a request and an optional interface is not available, this class is cached without that interface.
2. A future request comes in, and reuses the cache, but since this may be via a different code path it's possible that the interface is now available. Yet because the class was cached, this won't implement the interface now even though it's available at this point in time.

Reply:
Thanks @nielsdos, this is (so far) the hardest part for me, I've spent a lot of time with segfaults on this.

I moved the line as you suggested and the tests are passing now, but considering what you said, I don't understand why. If it's only applicable to non-immutable and non-cached classes, shouldn't it be more like this to make it work?

```php
if (ce->ce_flags & ZEND_ACC_IMMUTABLE) {
	// ...
} else if (ce->ce_flags & ZEND_ACC_FILE_CACHED) {
	// ...
} else {
	/* Only applicable for non-immutable and non-cached ce`s. */
	ce->num_interfaces = num_implementable_interfaces;
}
```

Or is the `ce` fine to be updated after loading? Tbh I don't even understand what is an immutable class. Is that an internal feature? Or am I forgetting some feature of PHP?

Reply:
The concept of an immutable class entry is an implementation detail, not a feature exposed to userland PHP.
An immutable class entry is one that lives in the shared memory of opcache.

When `zend_lazy_class_load` is called within the two if checks you referenced, it will allocate a new `ce` instance that _may_ be modified:

https://github.com/php/php-src/blob/789627612a487a60dfd8ee66c09d84090e22a382/Zend/zend_inheritance.c#L3324-L3328